packages/fxa-auth-server/lib/oauth/grant.js

```diff
@@ -0,0 +1,258 @@
+    token_type: access.type,
+    scope: access.scope.toString(),
+  };
+  result.expires_in =
+    grant.ttl || Math.floor((access.expiresAt - Date.now()) / 1000);
+  if (grant.authAt) {
+    result.auth_at = grant.authAt;
+  }
+  if (grant.keysJwe) {
+    result.keys_jwe = grant.keysJwe;
+  }
+  // Maybe also generate a refreshToken?
+  if (grant.offline) {
+    const refresh = await db.generateRefreshToken(grant);
+    result.refresh_token = refresh.token.toString('hex');
+  }
+  // Maybe also generate an idToken?
+  if (grant.scope && grant.scope.contains(SCOPE_OPENID)) {
+    result.id_token = await generateIdToken(grant, result.access_token);
+  }
+
+  if (grant.scope && grant.scope.contains(OAUTH_SCOPE_SESSION_TOKEN)) {
+    result.session_token_id =
+      grant.sessionTokenId && grant.sessionTokenId.toString('hex');
+  }
+
+  amplitude('token.created', {
+    service: hex(grant.clientId),
+    uid: hex(grant.userId),
+  });
+
+  return result;
+};
+
+async function generateIdToken(grant, accessToken) {
+  var now = Math.floor(Date.now() / 1000);
+  const clientId = hex(grant.clientId);
+  // The IETF spec for `aud` refers to https://openid.net/specs/openid-connect-core-1_0.html#IDToken
+  // > REQUIRED. Audience(s) that this ID Token is intended for. It MUST contain the
+  // > OAuth 2.0 client_id of the Relying Party as an audience value. It MAY also contain
+  // > identifiers for other audiences. In the general case, the aud value is an array of
+  // > case sensitive strings. In the common special case when there is one audience, the
+  // > aud value MAY be a single case sensitive string.
+  const audience = grant.resource ? [clientId, grant.resource] : clientId;
+
+  const claims = {
+    sub: await sub(grant.userId, grant.clientId, grant.ppidSeed),
+    aud: audience,
+    //iss set in jwt.sign
+    iat: now,
+    exp: now + ID_TOKEN_EXPIRATION,
+    at_hash: util.generateTokenHash(accessToken),
+  };
+  if (grant.amr) {
+    claims.amr = grant.amr;
+  }
+  if (grant.aal) {
+    claims['fxa-aal'] = grant.aal;
+    claims.acr = 'AAL' + grant.aal;
+  }
+
+  return jwt.sign(claims);
+}
+
+exports.generateAccessToken = async function generateAccessToken(grant) {
+  const accessToken = await db.generateAccessToken(grant);
+  if (
+    !JWT_ACCESS_TOKENS_ENABLED ||
+    !JWT_ACCESS_TOKENS_CLIENT_IDS.has(hex(grant.clientId).toLowerCase())
+  ) {
+    // return the old style access token if JWT access tokens are
+    // not globally enabled or if not enabled for the given clientId.
+    return accessToken;
+  }
+
+  if (grant.scope.contains('profile:subscriptions')) {
+    const capabilities = await determineClientVisibleSubscriptionCapabilities(
+      subConfig,
+      {},
+      authdb,
+      hex(grant.userId),
+      grant.clientId
+    );
+    // To avoid mutating the input grant, create a
+    // copy and add the new property there.
+    grant = {
+      ...grant,
+      'fxa-subscriptions': capabilities,
+    };
+  }
```

Comment:
In oauth-server this makes a request to the auth-server. Instead of making an request to itself I refactored it to get the data directly. One bit of hackery here is that `authdb` from up on line 55-58 that gets set by `lib/routes/index.js`. There's probably a nicer way to do it but it'd require a bigger change.